File: forward_price.py
import math


class ForwardPrice:
    def commodities(self, P, t, r, s, i):
        return P * (1 + r / 100 * t / 12) + s * t / 12 + i * t / 12

    def stock(self, P, r, d, t):
        return P * math.exp((r / 100 - d / 100) * t / 12)

    def bond(self, P, r, c, t):
        r_dec = r / 100
        c_dec = c / 100
        t_years = t / 12
        return P * (1 + (r_dec - c_dec) * t_years)

File: public/js/forward-price.js
const ForwardPrice = {
  commodities(P, t, r, s, i) {
    return P * (1 + (r / 100) * (t / 12)) + s * (t / 12) + i * (t / 12);
  },

  stock(P, r, d, t) {
    return P * Math.exp(((r / 100) - (d / 100)) * (t / 12));
  },

  bond(P, r, c, t) {
    const rDec = r / 100;
    const cDec = c / 100;
    const tYears = t / 12;
    return P * (1 + (rDec - cDec) * tYears);
  },
};

const CALCULATORS = {
  commodities: {
    title: "Physical Commodities",
    method: "commodities",
    fields: [
      ["P", "Commodity price"],
      ["t", "Time to maturity (months)"],
      ["r", "Interest rate (%)"],
      ["s", "Storage cost"],
      ["i", "Annual insurance cost"],
    ],
  },
  stock: {
    title: "Stock",
    method: "stock",
    fields: [
      ["P", "Spot price"],
      ["r", "Risk-free rate (%)"],
      ["d", "Dividend yield (%)"],
      ["t", "Time to maturity (months)"],
    ],
  },
  bond: {
    title: "Bond",
    method: "bond",
    fields: [
      ["P", "Bond price"],
      ["r", "Interest rate (%)"],
      ["c", "Coupon rate (%)"],
      ["t", "Time to maturity (months)"],
    ],
  },
};
